struct_insert.go

```diff
@@ -65,28 +67,85 @@ func (db *DB) buildInsert(record interface{}) *StructInsert {
 	return si
 }
 
+// Whitelist saves columns to be inserted from struct
+// It adds columns to list each time it is called
+// whitelist should not include auto key tagged columns
+func (si *StructInsert) Whitelist(columns ...string) *StructInsert {
+	si.whiteList = append(si.whiteList, columns...)
+	return si
+}
+
+// WhitelistReset resets whiteList
+func (si *StructInsert) WhitelistReset() *StructInsert {
+	si.whiteList = nil
+	return si
+}
+
+// Blacklist saves columns not to be inserted from struct
+// It adds columns to list each time it is called. If a column defined in whitelist is
+// also given in black list than that column will be blacklisted.
+func (si *StructInsert) Blacklist(columns ...string) *StructInsert {
+	si.blackList = append(si.blackList, columns...)
+	return si
+}
+
+// BlacklistReset resets blacklist
+func (si *StructInsert) BlacklistReset() *StructInsert {
+	si.blackList = nil
+	return si
+}
+
 // Do executes the insert statement.
 //
-// The behaviour differs according to the adapter. If it implements the
+// The behavior differs according to the adapter. If it implements the
 // InsertReturningSuffixer interface it will use it and fill all auto fields
 // of the given struct. Otherwise it only fills the key with LastInsertId.
 //
-// With BulkInsert the behaviour changeq occording to the adapter, see
-// BulkInsert documentation for more informations.
+// With BulkInsert the behavior changeq according to the adapter, see
+// BulkInsert documentation for more information.
 func (si *StructInsert) Do() error {
 	if si.error != nil {
 		return si.error
 	}
 
 	// Columns names
-	columns := si.recordDescription.structMapping.GetNonAutoColumnsNames()
-	si.insertStatement = si.insertStatement.Columns(si.insertStatement.db.quoteAll(columns)...)
+	var columns []string
+	if len(si.whiteList) > 0 {
+		columns = si.whiteList
+	} else {
+		columns = si.recordDescription.structMapping.GetNonAutoColumnsNames()
+	}
+	// Filter black listed columns
+	i := 0
+	for _, c := range si.blackList {
+		i = 0
+		for _, a := range columns {
+			if a != c {
+				columns[i] = a
+				i++
+			}
+		}
+		columns = columns[:i]
+	}
 
+	hasWB := (len(si.whiteList) + len(si.blackList)) > 0
+	if !hasWB {
+		si.insertStatement = si.insertStatement.Columns(si.insertStatement.db.quoteAll(columns)...)
+	}
 	// Values
+	values := make([]interface{}, 0, len(columns))
 	len := si.recordDescription.len()
+	wbColsSet := false
 	for i := 0; i < len; i++ {
 		currentRecord := si.recordDescription.index(i)
-		values := si.recordDescription.structMapping.GetNonAutoFieldsValues(currentRecord)
+		if hasWB {
+			columns, values = si.recordDescription.structMapping.GetNonAutoFieldsValuesFiltered(currentRecord, columns)
+			if !wbColsSet { // order of old columns list and current values list may not be same so, set here:
```

Comment:
`wbColsSet` need to be set to true here.

You can add a test with BulkInsert to reveal the error (and ensure there will be no regression) : 

```
			Convey("Do executes the query using a blacklist", func() {
				dummies := make([]Dummy, 0, 2)
				dummy1 := Dummy{
					AText:           "Dummy1",
					AnotherText:     "Foo",
					AnInteger:       12345,
					ANullableString: sql.NullString{String: "Void", Valid: true},
				}
				dummy2 := Dummy{
					AText:           "Dummy2",
					AnotherText:     "Foo",
					AnInteger:       12345,
					ANullableString: sql.NullString{String: "Void", Valid: true},
				}
				dummies = append(dummies, dummy1, dummy2)
				err := db.BulkInsert(&dummies).Blacklist("a_nullable_string").Do()
				So(err, ShouldBeNil)
			})
```

Reply:
Thats right, it is a footprint of my struggle to preserve to match `params` and `values`. First tried a map for columns and values, that failed because of random ordering of maps. Later used slices for `columns` and `values`.
Corrected. Thanks